File: nodejs/src/examples/02_mix_comparison.js
#!/usr/bin/env node
/**
 * 02_mix_comparison.js - Compare Two Mixes with the Tonn API
 *
 * This example demonstrates how to compare two audio files to identify
 * differences in loudness, dynamics, stereo field, and tonal profile.
 *
 * Usage:
 *   node 02_mix_comparison.js <file_a> <file_b> <musical_style> [--is-master]
 *
 * Example:
 *   node 02_mix_comparison.js my_mix.wav reference.wav POP
 */

import { existsSync } from 'fs';
import { TonnClient } from '../shared/client.js';

// ANSI colors
const colors = {
  red: '\x1b[91m',
  green: '\x1b[92m',
  yellow: '\x1b[93m',
  reset: '\x1b[0m',
  bold: '\x1b[1m'
};

/**
 * Analyze a single mix.
 */
async function analyzeMix(client, audioUrl, musicalStyle, isMaster) {
  const payload = {
    mixDiagnosisData: {
      audioFileLocation: audioUrl,
      musicalStyle,
      isMaster
    }
  };

  const response = await client.post('/mixanalysis', payload);
  return response?.mixDiagnosisResults || {};
}

/**
 * Extract production metrics from diagnosis.
 */
function extractMetrics(diagnosis) {
  const payload = diagnosis.payload || {};
  const keys = [
    'bit_depth', 'clipping', 'if_master_drc', 'if_master_loudness',
    'integrated_loudness_lufs', 'mono_compatible', 'peak_loudness_dbfs',
    'phase_issues', 'sample_rate', 'stereo_field'
  ];
  const metrics = {};
  for (const key of keys) {
    metrics[key] = payload[key] ?? 'N/A';
  }
  return metrics;
}

/**
 * Compare two values and return formatted output.
 */
function compareValue(key, valA, valB) {
  const thresholds = {
    integrated_loudness_lufs: 1.0,
    peak_loudness_dbfs: 0.5,
    bit_depth: 0
  };

  // Try numeric comparison
  const numA = parseFloat(valA);
  const numB = parseFloat(valB);

  if (!isNaN(numA) && !isNaN(numB)) {
    const diff = Math.abs(numA - numB);
    const threshold = thresholds[key] || 0;

    if (threshold && diff > threshold) {
      return {
        fmtA: `${colors.red}${valA}${colors.reset}`,
        fmtB: `${colors.red}${valB}${colors.reset}`,
        interpretation: `Difference of ${diff.toFixed(2)} exceeds threshold`
      };
    } else {
      return {
        fmtA: `${colors.green}${valA}${colors.reset}`,
        fmtB: `${colors.green}${valB}${colors.reset}`,
        interpretation: diff > 0 ? 'Within acceptable range' : 'Identical'
      };
    }
  }

  // String comparison
  if (String(valA) === String(valB)) {
    return {
      fmtA: `${colors.green}${valA}${colors.reset}`,
      fmtB: `${colors.green}${valB}${colors.reset}`,
      interpretation: 'Identical'
    };
  } else {
    return {
      fmtA: `${colors.yellow}${valA}${colors.reset}`,
      fmtB: `${colors.yellow}${valB}${colors.reset}`,
      interpretation: 'Values differ'
    };
  }
}

/**
 * Print comparison of two dictionaries.
 */
function compareDicts(dictA, dictB, title) {
  console.log(`\n${colors.bold}=== ${title} ===${colors.reset}\n`);

  const allKeys = [...new Set([...Object.keys(dictA), ...Object.keys(dictB)])].sort();

  for (const key of allKeys) {
    const valA = dictA[key] ?? 'N/A';
    const valB = dictB[key] ?? 'N/A';
    const { fmtA, fmtB, interpretation } = compareValue(key, valA, valB);

    console.log(`${key}:`);
    console.log(`  File A: ${fmtA}`);
    console.log(`  File B: ${fmtB}`);
    console.log(`  → ${interpretation}\n`);
  }
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length < 3 || args.includes('--help')) {
    console.log(`
Usage: node 02_mix_comparison.js <file_a> <file_b> <musical_style> [--is-master]

Arguments:
  file_a         First audio file to compare
  file_b         Second audio file to compare
  musical_style  Musical style (e.g., ROCK, POP)
  --is-master    Flag if comparing mastered tracks

Examples:
  node 02_mix_comparison.js my_mix.wav reference.wav POP
  node 02_mix_comparison.js v1.mp3 v2.mp3 ROCK --is-master
    `);
    process.exit(args.includes('--help') ? 0 : 1);
  }

  const fileA = args[0];
  const fileB = args[1];
  const musicalStyle = args[2].toUpperCase();
  const isMaster = args.includes('--is-master');

  // Validate files
  for (const filepath of [fileA, fileB]) {
    if (!existsSync(filepath)) {
      console.error(`Error: File not found: ${filepath}`);
      process.exit(1);
    }
  }

  const client = new TonnClient();
  const readableUrls = {};
  const results = {};

  // Upload both files
  for (const [label, filepath] of [['A', fileA], ['B', fileB]]) {
    console.log(`\n📤 Uploading File ${label}: ${filepath}...`);
    const url = await client.uploadLocalFile(filepath);
    if (!url) {
      console.error(`Failed to upload ${filepath}. Exiting.`);
      process.exit(1);
    }
    readableUrls[label] = url;
  }

  // Analyze both files
  console.log('\n🔍 Analyzing both files...');
  for (const [label, url] of Object.entries(readableUrls)) {
    console.log(`  Analyzing File ${label}...`);
    const diagnosis = await analyzeMix(client, url, musicalStyle, isMaster);
    if (!diagnosis || Object.keys(diagnosis).length === 0) {
      console.error(`Analysis failed for File ${label}`);
      process.exit(1);
    }
    results[label] = diagnosis;
  }

  // Compare results
  console.log('\n' + '='.repeat(60));
  console.log(`${colors.bold}MIX COMPARISON RESULTS${colors.reset}`);
  console.log('='.repeat(60));

  const metricsA = extractMetrics(results.A);
  const metricsB = extractMetrics(results.B);
  compareDicts(metricsA, metricsB, 'Production Metrics');

  const tonalA = results.A.payload?.tonal_profile || {};
  const tonalB = results.B.payload?.tonal_profile || {};
  if (Object.keys(tonalA).length || Object.keys(tonalB).length) {
    compareDicts(tonalA, tonalB, 'Tonal Profile');
  }

  console.log('='.repeat(60));
  console.log('Comparison complete!');
}

main().catch(console.error);

File: python/examples/02_mix_comparison.py
#!/usr/bin/env python3
"""
02_mix_comparison.py - Compare Two Mixes with the Tonn API

This example demonstrates how to compare two audio files (mixes or masters)
to identify differences in loudness, dynamics, stereo field, and tonal profile.

Use cases:
- Compare your mix to a reference track
- Compare two versions of the same mix
- A/B test different mastering approaches

Usage:
    python 02_mix_comparison.py <file_a> <file_b> <musical_style> [--is-master]

Example:
    python 02_mix_comparison.py my_mix.wav reference.wav POP
    python 02_mix_comparison.py v1.wav v2.wav ROCK --is-master
"""

import sys
import os
import argparse

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from shared import TonnClient


# ANSI colors for terminal output
class Colors:
    RED = "\033[91m"
    GREEN = "\033[92m"
    YELLOW = "\033[93m"
    RESET = "\033[0m"
    BOLD = "\033[1m"


def analyze_mix(client: TonnClient, audio_url: str, musical_style: str, is_master: bool) -> dict:
    """Analyze a single mix and return the diagnosis results."""
    payload = {
        "mixDiagnosisData": {
            "audioFileLocation": audio_url,
            "musicalStyle": musical_style,
            "isMaster": is_master
        }
    }
    
    response = client.post("/mixanalysis", payload)
    if response and "mixDiagnosisResults" in response:
        return response["mixDiagnosisResults"]
    return {}


def extract_metrics(diagnosis: dict) -> dict:
    """Extract production metrics from diagnosis payload."""
    payload = diagnosis.get("payload", {})
    keys = [
        "bit_depth", "clipping", "if_master_drc", "if_master_loudness",
        "integrated_loudness_lufs", "mono_compatible", "peak_loudness_dbfs",
        "phase_issues", "sample_rate", "stereo_field"
    ]
    return {key: payload.get(key, "N/A") for key in keys}


def extract_tonal_profile(diagnosis: dict) -> dict:
    """Extract tonal profile from diagnosis payload."""
    return diagnosis.get("payload", {}).get("tonal_profile", {})


def compare_value(key: str, val_a, val_b) -> tuple:
    """Compare two values and return formatted strings with interpretation."""
    # Thresholds for numeric comparisons
    thresholds = {
        "integrated_loudness_lufs": 1.0,
        "peak_loudness_dbfs": 0.5,
        "bit_depth": 0
    }
    
    # Try numeric comparison
    try:
        num_a, num_b = float(val_a), float(val_b)
        diff = abs(num_a - num_b)
        threshold = thresholds.get(key, 0.0)
        
        if threshold and diff > threshold:
            fmt_a = f"{Colors.RED}{val_a}{Colors.RESET}"
            fmt_b = f"{Colors.RED}{val_b}{Colors.RESET}"
            interpretation = f"Difference of {diff:.2f} exceeds threshold"
        else:
            fmt_a = f"{Colors.GREEN}{val_a}{Colors.RESET}"
            fmt_b = f"{Colors.GREEN}{val_b}{Colors.RESET}"
            interpretation = "Within acceptable range" if diff > 0 else "Identical"
        
        return fmt_a, fmt_b, interpretation
        
    except (ValueError, TypeError):
        # Non-numeric comparison
        if str(val_a) == str(val_b):
            fmt_a = f"{Colors.GREEN}{val_a}{Colors.RESET}"
            fmt_b = f"{Colors.GREEN}{val_b}{Colors.RESET}"
            interpretation = "Identical"
        else:
            fmt_a = f"{Colors.YELLOW}{val_a}{Colors.RESET}"
            fmt_b = f"{Colors.YELLOW}{val_b}{Colors.RESET}"
            interpretation = "Values differ"
        
        return fmt_a, fmt_b, interpretation


def compare_dicts(dict_a: dict, dict_b: dict, title: str) -> None:
    """Print a side-by-side comparison of two dictionaries."""
    print(f"\n{Colors.BOLD}=== {title} ==={Colors.RESET}\n")
    
    all_keys = sorted(set(dict_a.keys()) | set(dict_b.keys()))
    
    for key in all_keys:
        val_a = dict_a.get(key, "N/A")
        val_b = dict_b.get(key, "N/A")
        fmt_a, fmt_b, interpretation = compare_value(key, val_a, val_b)
        
        print(f"{key}:")
        print(f"  File A: {fmt_a}")
        print(f"  File B: {fmt_b}")
        print(f"  → {interpretation}\n")


def main():
    parser = argparse.ArgumentParser(
        description="Compare two audio mixes using the Tonn API",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python 02_mix_comparison.py my_mix.wav reference.wav POP
  python 02_mix_comparison.py v1.mp3 v2.mp3 ROCK --is-master
        """
    )
    parser.add_argument("file_a", help="First audio file to compare")
    parser.add_argument("file_b", help="Second audio file to compare")
    parser.add_argument("musical_style", help="Musical style (e.g., ROCK, POP)")
    parser.add_argument("--is-master", action="store_true",
                       help="Flag if comparing mastered tracks")

    args = parser.parse_args()

    # Validate files exist
    for filepath in [args.file_a, args.file_b]:
        if not os.path.isfile(filepath):
            print(f"Error: File not found: {filepath}")
            sys.exit(1)

    client = TonnClient()
    results = {}
    readable_urls = {}

    # Upload both files
    for label, filepath in [("A", args.file_a), ("B", args.file_b)]:
        print(f"\n📤 Uploading File {label}: {os.path.basename(filepath)}...")
        url = client.upload_local_file(filepath)
        if not url:
            print(f"Failed to upload {filepath}. Exiting.")
            sys.exit(1)
        readable_urls[label] = url

    # Analyze both files
    print(f"\n🔍 Analyzing both files...")
    for label, url in readable_urls.items():
        print(f"  Analyzing File {label}...")
        diagnosis = analyze_mix(client, url, args.musical_style.upper(), args.is_master)
        if not diagnosis:
            print(f"Analysis failed for File {label}")
            sys.exit(1)
        results[label] = diagnosis

    # Compare results
    print("\n" + "=" * 60)
    print(f"{Colors.BOLD}MIX COMPARISON RESULTS{Colors.RESET}")
    print("=" * 60)

    # Production metrics comparison
    metrics_a = extract_metrics(results["A"])
    metrics_b = extract_metrics(results["B"])
    compare_dicts(metrics_a, metrics_b, "Production Metrics")

    # Tonal profile comparison
    tonal_a = extract_tonal_profile(results["A"])
    tonal_b = extract_tonal_profile(results["B"])
    if tonal_a or tonal_b:
        compare_dicts(tonal_a, tonal_b, "Tonal Profile")

    print("=" * 60)
    print("Comparison complete!")


if __name__ == "__main__":
    main()

